### ISSUE TITLE
Relabel rules: a label that is deleted then set will read back as empty

### ISSUE THREAD

**Issue by hawran** [prometheus/prometheus#12283]
### What did you do?

A snippet from `relabel_configs:`
```
  - source_labels: [__meta_kubernetes_pod_annotation_XXX_metrics_port]
    action: replace
    regex: (.+)
    replacement: metrics
    target_label: __meta_kubernetes_pod_container_port_name
  - source_labels: [__meta_kubernetes_pod_container_port_name]
    regex: ^metrics$
    action: keep
```

### What did you expect to see?

All metrics from a pod with `XXX.metrics.port` annotation are scraped.
(As worked up to `2.42.0` version (inclusive).)

### What did you see instead? Under which circumstances?

No metrics from a pod.
After upgrade to `2.43.0`.

### Logs

```text
Please, pay attention to:

https://github.com/prometheus/prometheus/blob/main/model/labels/labels.go#L492
https://github.com/prometheus/prometheus/blob/main/model/labels/labels.go#L519
https://github.com/prometheus/prometheus/blob/main/model/labels/labels.go#L535
```

**Comment by hawran** [comment]
### Prometheus version
`2.43.0`
(I am sorry.)

**Comment by bboreham** [comment]
What is the "error" from your title?

Also, I did not understand the "Please, pay attention to:" part.  Please explain again in a different way.

**Comment by roidelapluie** [comment]
Can you provide the problematic set of labels that are dropped but should not, with the complete relabeling rules?

**Comment by hawran** [comment]
Hi, I'm sorry for being so monosyllabic.
Further info is as follows:

> What is the "error" from your title?

Well, I'd presumed when all metrics were gone after the upgrade and the `Target Labels` read `Dropped` within the `Service Discovery` view (Prometheus) there could be something wrong with relabeling.
 
> Also, I did not understand the "Please, pay attention to:" part. Please explain again in a different way.

There was a small misunderstanding with a colleague of mine over here, I C&P wrong links, sorry for that.

FMPOV I can see some inconsistency at
https://github.com/prometheus/prometheus/blame/release-2.43/model/labels/labels.go#L592
vs.
https://github.com/prometheus/prometheus/blame/release-2.43/model/labels/labels.go#L534
What if there's a same label within both `add` and `del` arrays?

TBH, the `Set()` function
https://github.com/prometheus/prometheus/blame/release-2.43/model/labels/labels.go#L517
seems suspicious to me as well - it ignores the `del` array and just sets the `add` array, but it might be just my ignorance.

**Comment by hawran** [comment]
Real use cases for the snippet
```
  - source_labels: [__meta_kubernetes_pod_annotation_XXX_metrics_port]
    action: replace
    regex: (.+)
    replacement: metrics
    target_label: __meta_kubernetes_pod_container_port_name
  - source_labels: [__meta_kubernetes_pod_container_port_name]
    regex: ^metrics$
    action: keep
```
as follows:

When there is the Discovered Label `__meta_kubernetes_pod_annotation_XXX_metrics_port="9091"` set
`2.42.0`: metrics are scraped
`2.43.0`: no metrics.

When there are both Discovered Labels `__meta_kubernetes_pod_annotation_XXX_metrics_port="9091"` and `__meta_kubernetes_pod_container_port_name="metrics"` set,
`2.42.0`: metrics are scraped
`2.43.0`: metrics are scraped.

When there is the Discovered Label `__meta_kubernetes_pod_container_port_name="metrics"` set,
`2.42.0`: metrics are scraped
`2.43.0`: metrics are scraped.

**Comment by hawran** [comment]
As for
> ... MPOV I can see some inconsistency at...

[metioned above](https://github.com/prometheus/prometheus/issues/12283#issuecomment-1519742169), we've tested steps:

1. `Del("LABEL")`
2. `Set("LABEL")`

from `Get("LABEL")` we've got nothing, from `Labels(...)` we've got the label `LABEL`.

**Comment by bboreham** [comment]
OK, now I see, if you call `Del` followed by `Set` of the same label name, then Builder will sometimes not return the value.
That was unintended.

**Comment by bboreham** [comment]
At #12322 I tried to create the problem based on the relabeling rules you quoted, but I had to add something at the beginning to do the initial `Del()`.  If you can see what I missed from the context of your rules, or your study of the code, please let me know.

**Comment by bboreham** [comment]
@hawran are you able to test with the new 2.44 release candidate? https://github.com/prometheus/prometheus/releases/tag/v2.44.0-rc.1 ?

**Comment by hawran** [comment]
> At #12322 I tried to create the problem based on the relabeling rules you quoted, but I had to add something at the beginning to do the initial `Del()`. If you can see what I missed from the context of your rules, or your study of the code, please let me know.

Well, I've pasted just a little snippet from somewhere in the middle of a much bigger `relabel_configs` section.
The `Del("__meta_kubernetes_pod_container_port_name")` function is called internally by prometheus.

**Comment by hawran** [comment]
> @hawran are you able to test with the new 2.44 release candidate? https://github.com/prometheus/prometheus/releases/tag/v2.44.0-rc.1 ?

We've upgraded to `2.43.1` and it seems to be OK now, thank you.

**Comment by bboreham** [comment]
> The `Del("__meta_kubernetes_pod_container_port_name")` function is called internally by prometheus.

Do you know any more about how this comes about?  I just want to make sure my test case is accurate.

> We've upgraded to 2.43.1 and it seems to be OK now, thank you.

That's great.  I think we will leave this issue open until the fix is merged to `main`.

**Comment by bboreham** [comment]
Would you mind if I re-titled this issue to "Relabel rules: a label that is deleted then set will read back as empty" ?
To me, this is a clearer description of the bug, and should make it easier for others to find in future.

**Comment by hawran** [comment]
> Would you mind if I re-titled this issue to "Relabel rules: a label that is deleted then set will read back as empty" ? To me, this is a clearer description of the bug, and should make it easier for others to find in future.

No problem, feel free to change the title as you need.

**Comment by bboreham** [comment]
Fixed by #12322
